dali/imgcodec/parsers/webp.cc

```diff
@@ -12,19 +12,118 @@
+  ssize_t length = stream->Size();
+
+  // Skipping 12 bytes of the RIFF header chunk
+  // "RIFF" (4 bytes)
+  // File Size (4 bytes)
+  // "WEBP" (4 bytes)
+  stream->SeekRead(12, SEEK_CUR);
+
   ImageInfo info;
-  // info.shape = {h, w, c};   // TODO(janton)
+  const auto header = stream->ReadOne<std::array<uint8_t, 4>>();
+  if (is_pattern_matching(header.data(), "VP8 ")) {
+    // Simple file format (lossy)
+    // https://datatracker.ietf.org/doc/html/rfc6386#section-9.1
+    // Skipping chunk size (4 bytes) and frame tag (3 bytes)
+    stream->SeekRead(7, SEEK_CUR);
+
+    // Verify sync code
+    const auto sync_code = stream->ReadOne<std::array<uint8_t, 3>>();
+    if (!is_valid_sync_code_lossy(sync_code.data())) {
+      DALI_FAIL("Sync code 157 1 42 not found at expected position. Found " +
+                std::to_string(sync_code[0]) + " " +
+                std::to_string(sync_code[1]) + " " +
+                std::to_string(sync_code[2]) + " instead.");
+    }
+
+    const int w = ReadValueLE<uint16_t>(*stream) & 0x3FFF;
+    const int h = ReadValueLE<uint16_t>(*stream) & 0x3FFF;
+    info.shape = {h, w, 3};
+  } else if (is_pattern_matching(header.data(), "VP8L")) {
+    // Simple file format (lossless)
+    // https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#2_riff_header
+    // Skipping number of bytes in the lossless stream (4 bytes)
+    stream->SeekRead(4, SEEK_CUR);
+
+    // Verify the signature byte
+    const auto sync_code = stream->ReadOne<uint8_t>();
+    if (sync_code != SIGNATURE_BYTE_LOSTLESS) {
+        DALI_FAIL("Sync code 47 not found at expected position. Found " +
+                  std::to_string(sync_code) + " instead.");
+    }
+
+    // VP8L shape information starts after the sync code
+    const auto features = ReadValueLE<uint32_t>(*stream);
+    const int w = (features & 0x00003FFF) + 1;
+    const int h = ((features & 0x0FFFC000) >> 14) + 1;
+    const int alpha = (features & 0x10000000) >> 28;
+
+    // VP8L always uses RGBA
+    info.shape = {h, w, 3 + alpha};
+  } else if (is_pattern_matching(header.data(), "VP8X")) {
+    // Extended file format
+    DALI_FAIL("WebP extended file format is not supported.");
+  } else {
+    DALI_FAIL("Unrecognized WebP header: " +
+              std::to_string(header[0]) + " " +
+              std::to_string(header[1]) + " " +
+              std::to_string(header[2]) + " " +
+              std::to_string(header[3]));
+  }
+
   return info;
 }
 
 bool WebpParser::CanParse(ImageSource *encoded) const {
-  return false;  // TODO(janton)
+  // Sync code in lossy version ends at byte 25
+  uint8_t header[26];
+  if (!ReadHeader(header, encoded, sizeof(header)))
+    return false;
+
+  // Check the RIFF header
+  if (!is_pattern_matching(header, "RIFF"))
+    return false;
+  // 4 bytes of file size skipped
+  if (!is_pattern_matching(header + 8, "WEBP"))
+    return false;
+
+  if (is_pattern_matching(header + 12, "VP8 ")) {
```

Comment:
After reading the first part that handles the seeks basically "under the hood", this kinda feels like repeating the old, worse patterns. But I have no idea how to remove the repetition on make it better, I guess this is so small that we can live with it.